### DISCUSSION TITLE
RFC: Custom Controller Validation (draft)

### DISCUSSION THREAD

**Discussion by matthewmueller** [livebud/bud#15] [Proposals]
For more complex validation, you can define a `Valid` method on an input type:

```go
type Input struct {
  Name  string `json:"name"`
  Email string `json:"email"`
}

func (i *Input) Valid() error {
  switch {
    case len(i.Name) < 3:
      return fmt.Errorf("%s is not long enough", i.Name)
    case !strings.Contains(i.Email, "@"):
      return fmt.Errorf("%s is not a valid email", i.Email)	
    default:
      return nil
  }
}

func (c *Controller) Create(in *Input) {}
```

If validation fails, you get a 400 Bad Request response with your error message:

```go
{
  "error": "name is not long enough"
}
```

Validation also works with other input types:

```go
type Email string

func (email Email) Valid() error {
  if !strings.Contains(string(email), "@") {
    return errors.New("is not valid")
  }
  return nil
}

func (c *Controller) Create(email Email) {}
```

If validation fails, you get a `400 Bad Request` response. Since email is a key within the request map, we can provide more information in the error message:

```go
{
  "error": "email is not valid",
  "fields": {
    "email": "is not valid"
  }
}
```

## Open Questions

- What are your thoughts on calling it `Validate` vs `Valid`? IMO `Validate` sounds more like an action that would return an error, but `Valid` is shorter.

**Comment by matthewmueller** [comment]
There's an existing discussion about using validator here: https://github.com/livebud/bud/issues/171

**Comment by jfmario** [comment]
This would be a good solution I think.
